oxide_core/quaternion.rs

```diff
@@ -0,0 +1,169 @@
+use crate::vector::vector3::*;
+use crate::vector::*;
+use std::ops;
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct Quaternion {
+  pub x: f32,
+  pub y: f32,
+  pub z: f32,
+  pub w: f32,
+}
+
+pub trait VectorPart {
+  fn get_vector_part(&self) -> Vector3;
+}
+
+impl VectorPart for Quaternion {
+  fn get_vector_part(&self) -> Vector3 {
+    Vector3 {
+      x: self.x,
+      y: self.y,
+      z: self.z,
+    }
+  }
+}
+
+pub trait ScalarPart {
+  fn get_scalar_part(&self) -> f32;
+}
+
+impl ScalarPart for Quaternion {
+  fn get_scalar_part(&self) -> f32 {
+    self.w
+  }
+}
+
+pub fn id() -> Quaternion {
+  let one = 1.0f32;
+  let zero = 0.0f32;
+
+  Quaternion {
+    x: zero,
+    y: zero,
+    z: zero,
+    w: one,
+  }
+}
```

Comment:
What does the function name `id()` mean?

Reply:
It means identity quaternion. Should I have to rename it for making it clear?

Reply:
I think that this function should be removed. Instead, I suggest the function `set_identity()` or `make_identity()` like this:
```C++
template <typename T>
void Quaternion<T>::SetIdentity()
{
    Set(1, 0, 0, 0);
}
```
or
```C++
template <typename T>
Quaternion<T> Quaternion<T>::MakeIdentity()
{
    return Quaternion{};
}
```

Reply:
I'll implement both.

Reply:
✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)
Test implemented., [`b2c3eee` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/b2c3eee61212a1ec5c14f8b1960c285eb70639e8)